### PR TITLE
xdg-desktop-portal: fix build when doCheck = false

### LINKED ISSUE #275863: xdg-desktop-portal: build failure when `doCheck = false`
### Steps To Reproduce

When building `xdg-desktop-portal` with `doCheck = false`, the build fails with the error `Program 'pytest-3 pytest' not found or not executable`.

### Build log

Log excerpt:

```
(...)
Program /nix/store/7753941pnr2kii6csvnj445r84rbv92f-glib-2.78.1-dev/bin/gdbus-codegen found: YES (/nix/store/7753941pnr2kii6csvnj445r84rbv92f-glib-2.78.1-dev/bin/gdbus-codegen)
Configuring xdg-document-portal.service using configuration
Configuring org.freedesktop.portal.Documents.service using configuration
Configuring org.freedesktop.impl.portal.PermissionStore.service using configuration
Configuring test.portal using configuration
Configuring limited.portal using configuration
Configuring org.freedesktop.portal.Documents.service using configuration
Configuring org.freedesktop.impl.portal.PermissionStore.service using configuration
Program pytest-3 pytest found: NO

tests/meson.build:268:9: ERROR: Program 'pytest-3 pytest' not found or not executable

A full log can be found at /build/source/build/meson-logs/meson-log.txt
```

### Additional context

This is when trying to build NixOS 23.11.

### Notify maintainers

cc @jtojnar @pks-t @bobby285271 

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by someplaceguy**
## Description of changes

Fixes #275863

cc @jtojnar @pks-t @bobby285271

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by jtojnar — CHANGES_REQUESTED** [review]
This is expected, `doCheck` disabled running tests but Meson has no idea they will not run. If you want to make `doCheck = false` work, you should also disable building/configuring tests.

**Comment by someplaceguy** [comment]
@jtojnar How do I disable building tests?

I know nothing about meson and the `meson` section in the `nixpkgs` manual doesn't mention anything about disabling the building of tests.

Would removing the `"-Dinstalled-tests=true"` flag when `doCheck = false` do that or is something else needed?

**Comment by eclairevoyant** [comment]
using `lib.optionals` on the two meson flags related to tests would be a start, yes.

**Comment by someplaceguy** [comment]
It turned out that I needed to add another flag instead...

File: pkgs/development/libraries/xdg-desktop-portal/default.nix [hunk 1]
```diff
@@ -110,6 +110,8 @@ stdenv.mkDerivation (finalAttrs: {
     "-Dinstalled_test_prefix=${placeholder "installedTests"}"
   ] ++ lib.optionals (!enableGeoLocation) [
     "-Dgeoclue=disabled"
+  ] ++ lib.optionals (!doCheck) [
```
Comment (eclairevoyant) [inline pkgs/development/libraries/xdg-desktop-portal/default.nix:113]:
```suggestion
  ] ++ lib.optionals (!finalAttrs.doCheck) [
```

File: pkgs/development/libraries/xdg-desktop-portal/default.nix [hunk 2]
```diff
@@ -29,7 +29,7 @@
 , enableGeoLocation ? true
 }:
 
-stdenv.mkDerivation (finalAttrs: {
+stdenv.mkDerivation (finalAttrs: rec {
```
Comment (eclairevoyant) [inline pkgs/development/libraries/xdg-desktop-portal/default.nix:32]:
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

**Review by eclairevoyant — APPROVED** [review]
Builds with the override now, approving.

**Comment by eclairevoyant** [comment]
Result of `nixpkgs-review pr 275867` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by eclairevoyant** [comment]
@jtojnar mind looking at this? changes were made and the previous review is currently blocking merge

**Comment by jtojnar** [comment]
Right, this should work but I am still not sure why we would want this.

Disabling tests is outside of regular use case IMO, and if a person is already overriding `doCheck`, they can just as well pass the appropriate flags, so I do not see the point of increasing the complexity of the expression.

Sorry, I should have been clearer about considering this something that should be done by the consumer.

**Comment by someplaceguy** [comment]
> Right, this should work but I am still not sure why we would want this.

There are at least two good reasons:

1. `xdg-desktop-portal`'s tests are unreliable (search for "timeout" inside the `tests/` directory and you will find dozens of matches). Any test with a timeout is guaranteed to fail if the build machine's load is high enough (which is what happened in my case) or if the machine is slow enough (e.g. when running under an emulator for another architecture).

2. Cross-compilation also disables `doCheck`

**Comment by Frontear** [comment]
> > Right, this should work but I am still not sure why we would want this.
> 
> There are at least two good reasons:
> 
> 1. `xdg-desktop-portal`'s tests are unreliable (search for "timeout" inside the `tests/` directory and you will find dozens of matches).
> 
> Any test with a timeout is _guaranteed_ to fail if the build machine's load is high enough (which is what happened in my case) or if the machine is slow enough (e.g. when running under an emulator for a different architecture).
> 
> Such test failures have nothing to do with the code that is being tested, the problem is in the tests themselves.
> 
> 2. Cross-compilation disables `doCheck`.

I've run into very similar issues myself for large-scale PRs (see #342961 and its history :p), and I think it would honestly be worth considering this a more prevalent issue in Nixpkgs. A lot of builds fail on Hydra constantly for this exact reason as seen in [#staging:nixos.org](https://matrix.to/#/#staging:nixos.org). We need a more standardized resolution to these types of problems, as disabling tests/increasing timeouts isn't a good all-cases fix.

**Comment by uninsane** [comment]
this was fixed by https://github.com/NixOS/nixpkgs/pull/342669

**Pull request closed without merging**